crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1838,6 +1842,8 @@ mod tests {
     #[test_case(Ty::LiteralString, Ty::BuiltinInstance("str"))]
     #[test_case(Ty::BytesLiteral("foo"), Ty::BuiltinInstance("bytes"))]
     #[test_case(Ty::IntLiteral(1), Ty::Union(vec![Ty::BuiltinInstance("int"), Ty::BuiltinInstance("str")]))]
+    #[test_case(Ty::Union(vec![Ty::BuiltinInstance("str"), Ty::BuiltinInstance("int")]), Ty::BuiltinInstance("object"))]
+    #[test_case(Ty::Union(vec![Ty::IntLiteral(1), Ty::IntLiteral(2)]), Ty::Union(vec![Ty::IntLiteral(1), Ty::IntLiteral(2), Ty::IntLiteral(3)]))]
```

Comment:
Unit tests for this are good, because this is a complex method, and it's good to have confidence that this specific part of the code is all working according to how we'd like it to work. I would like us to _also_ add inference tests using mdtest where possible, though, so that if we do a large refactor in the future and remove this method (and the tests for the method) entirely we'll be able to have confidence that we still retain the more sophisticated understanding of subtyping that this PR gives us.

I can see that adding inference tests using mdtest might be tricky for this _specific_ thing right now, though! So this is just a general point, not something that demands action on this PR specifically.